src/ocean/Ocean.C

```diff
@@ -2200,27 +2179,80 @@ void Ocean::setPar(std::string const &parName, double value)
 }
 
 //====================================================================
-void Ocean::setParameters(ParameterList pars)
+
+Teuchos::ParameterList
+Ocean::getDefaultInitParameters()
 {
-    std::string parName;
-    double parValue;
-    // This is similar to reading from HDF5
-    for (int par = 1; par <= _NPAR_; ++par)
-    {
-        parName  = THCM::Instance().int2par(par);
-        parValue = getPar(parName);
+    Teuchos::ParameterList result = getDefaultParameters();
+    result.sublist("THCM") = THCM::getDefaultInitParameters();
+    return result;
+}
 
-        // Overwrite continuation parameter and put it in THCM
-        try
-        {
-            parValue = pars->get(parName, parValue);
-        }
-        catch (EpetraExt::Exception &e)
-        {
-            e.Print();
-            continue;
-        }
+Teuchos::ParameterList
+Ocean::getDefaultParameters()
+{
+    Teuchos::ParameterList result;
+    result.get("Load salinity flux", false);
+    result.get("Save salinity flux", true);
+    result.get("Load temperature flux", false);
+    result.get("Save temperature flux", true);
 
-        setPar(parName, parValue);
-    }
+    result.get("Use legacy fort.3 output", false);
+    result.get("Use legacy fort.44 output", true);
+    result.get("Save column integral", false);
+    result.get("Max mask fixes", 5);
+
+    result.get("Analyze Jacobian", true);
+
+    result.get("Input file",  "ocean_input.h5");
+    result.get("Output file", "ocean_output.h5");
+    result.get("Save mask", true);
+    result.get("Load mask", true);
+
+    result.get("Load state", false);
+    result.get("Save state", true);
+    result.get("Save frequency", 0);
+    result.get("Store everything", false);
+
+    Teuchos::ParameterList& solverParams = result.sublist("Belos Solver");
+    solverParams.get("FGMRES iterations", 500);
+    solverParams.get("FGMRES tolerance", 1e-8);
+    solverParams.get("FGMRES restarts", 0);
+    solverParams.get("FGMRES output", 100);
+    solverParams.get("FGMRES explicit residual test", false);
+
+    result.sublist("THCM") = THCM::getDefaultParameters();
+
+    return result;
+}
+
+const Teuchos::ParameterList& Ocean::getParameters()
+{ return params_; }
+
+void Ocean::setParameters(Teuchos::ParameterList newParams)
```

Comment:
This should receive a reference to a parameterlist, not a copy. And then `validateParametersAndSetDefaults` should be called on that reference, not on a local copy. That way you can see in the parameterlist what gets used and what not (and when we properly document the parameters, that will be added too). The same should be done for the THCM parameterlist.

In `run_ocean`, if you do
```
    std::cout << *oceanParams << std::endl;
```
at the end, it should display
```
Load state = 0
Load salinity flux = 0
Save salinity flux = 1   [default]
Load temperature flux = 0   [default]
Save temperature flux = 1   [default]
Use legacy fort.3 output = 0   [default]
Use legacy fort.44 output = 1   [default]
Save column integral = 0   [default]
Max mask fixes = 5   [default]
Analyze Jacobian = 1   [default]
Input file = ocean_input.h5   [default]
Output file = ocean_output.h5   [default]
...
```
not
```
Load state = 0   [unused]
Load salinity flux = 0   [unused]
THCM -> 
 Problem Description = test   [unused]
 Global Bound xmin = 300   [unused]
...
```

You can use the `ocean` run directory and set
```
    <Parameter name="Global Grid-Size n" type="int" value="8"/>
    <Parameter name="Global Grid-Size m" type="int" value="8"/>
    <Parameter name="Global Grid-Size l" type="int" value="8"/>
```
if you want a quick example.

Reply:
Hmm...the reason for doing it this way was so that it would be valid to call `setParameters` with a subset of the parameters to update one or more parameters that `Ocean` is using.

If you do it your way you run into the problem calling `setParameters` with a parameter list that has a single parameter you want to update, then you reset all the other parameters back to default. You should still be able to figure out which parameters are used/defaulted by simply calling `getParameters()` and inspecting the results of that, no?

Reply:
hmm, I need to think about this for a bit, because that doesn't propagate the THCM usage back to Ocean, no. So I agree the constructors should be mutable references or `RCP`, but I need to think how to make that work with `setParameters` so that you can still update subsets of parameters.

Or, I guess the simpler way is to simply say that `setParameters` must always have *all* parameters and doesn't allow/do updates to subsets and then changing OMUSE to simply remember the previous parameters you set, which shouldn't be too difficult.

Reply:
What about you do

```
    params_.setParameters(newParams);
```
at the end of `setParameters` and then do
```
Teuchos::ParameterList result(params_);
```
at the start of `getDefaultParameters`?

Reply:
I think the correct solution is a compromise. I think the current validation is the right way to implement it, since your comment would actually make `getDefaultParameters` equal to `getParameters`.

But you're right that the parameter lists should be `RCP` pointers, and I think I finally got why they were like that. (On the other hand, refactoring them to use const references was helpful in sorting out the rest of the code, so that's good!)

Reply:
Ok, then maybe add an optional argument to `getDefault(Init)Parameters`? If it's set, you initialize result with that, if not then you do what is currently done.

Reply:
I don't think that's needed? In the new commit I've kept that code as is and it correctly tracks whether parameters use the default or not. I tried your example of having an `ocean_params.xml` that only sets the THCM dimensions and nothing else and when I print `oceanParams` at the end of `run_ocean` it correctly reports all variables as default, except the dimensions I set in THCM, which it reports as used.